cmd/k8s-operator/headscale/client_wrapper.go

```diff
@@ -0,0 +1,124 @@
+package headscale
```

Comment:
Why are we bringing a Headscale client into Tailscale?

I thought Headscale implemented Tailscale.

If Headscale has forked, that's great, but why isn't this in the Headscale repo as a Headscale Kubernetes Operator? 

I haven't been following Headscale closely. What am I missing context-wise?

Reply:
See https://github.com/tailscale/tailscale/pull/10223 for context.

Basically, Headscale only implements the VPN control plane that Tailscale Inc hosts at `https://controlplane.tailscale.com`. However, it does not implement the (management) API hosted by Tailscale Inc at `https://api.tailscale.com`.
The k8s operator needs the API to create auth keys and to clean up nodes for deleted services.

The package here (and maybe it could be named more aptly) provides a client for the Headscale (management) API, not a Headscale VPN client (which doesn't exist). More specifically, it wraps the Headscale API client library to implement the `tsClient` interface that was already defined in the operator.

Reply:
Hi @sysedwinistrator thank you for working on this. I understand wanting to get the operator to work with Headscale.

We will not add Headscale-specific code into the operator. The maintenance costs associated with having a client that essentially works with two different APIs would be too high. Additionally having to maintain compatibility with another API would stall our ability to make changes in the operator in the future. The features of tailscale API that we use now may well need to change in the future and we would compromise our ability to add new features and keep up with future changes made to the Tailscale API if we added a dependency on Headscale.

I am not sufficiently familiar with Headscale to understand how much work is needed to get Headscale working with the operator. I do see that folks Headscale-side say that this is out of scope (https://github.com/juanfont/headscale/issues/1202#issuecomment-1542350091, cc @kradalby ). Perhaps if other folks are willing to contribute they would change their mind (but this is not me making suggestions what the Headscale project that I am not involved with should do :) )